src/cata_tiles.cpp

```diff
@@ -889,15 +884,44 @@ void cata_tiles::draw( int destx, int desty, const tripoint &center, int width,
     auto vision_cache = g->u.get_vision_modes();
     nv_goggles_activated = vision_cache[NV_GOGGLES];
 
+    std::vector<tile_render_info> draw_points;
     for( y = min_y; y * dy < max_y * dy; y += dy) {
         for( x = min_x; x * dx < max_x * dx; x += dx) {
-            //if the render area is outside the visibility cache, default to the darkened tile
             if(!tile_iso && ((y < min_visible_y || y > max_visible_y) || (x < min_visible_x ||
                              x > max_visible_x))) {
                 apply_vision_effects(x, y, offscreen_type);
-            } else {
-                draw_single_tile( temp, ch.visibility_cache[x][y], cache );
+                continue;
             }
+
+            if( apply_vision_effects( x, y, g->m.get_visibility( ch.visibility_cache[x][y], cache ) ) ) {
+                const auto critter = g->critter_at( tripoint(x,y,center.z), true );
+                if( critter != nullptr && g->u.sees_with_infrared( *critter ) ) {
+                    //TODO defer drawing this until later when we know how tall
+                    //     the terrain/furniture under the creature is.
+                    draw_from_id_string( "infrared_creature", C_NONE, empty_string, x, y, 0, 0, LL_LIT, false );
+                }
+                continue;
+            }
+
+            int height_3d;
```

Comment:
aha! Good catch. It didn't occur to me that things would draw before the terrain draw calls initialize this. That probably caused a number of the problems.